File: transpiler/src/test/java/com/here/ivi/api/SmokeTest.java
/*
 * Copyright (C) 2017 HERE Global B.V. and its affiliate(s). All rights reserved.
 *
 * This software, including documentation, is protected by copyright controlled by
 * HERE Global B.V. All rights are reserved. Copying, including reproducing, storing,
 * adapting or translating, any or all of this material requires the prior written
 * consent of HERE Global B.V. This material also contains confidential information,
 * which may not be disclosed to others without prior written consent of HERE Global B.V.
 *
 */

package com.here.ivi.api;

import static org.junit.Assert.*;
import static org.junit.Assume.assumeFalse;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.spy;

import com.here.ivi.api.generator.common.GeneratedFile;
import com.here.ivi.api.generator.common.GeneratorSuite;
import java.io.File;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.*;
import java.util.List;
import java.util.stream.Collectors;
import org.eclipse.xtext.util.Files;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;
import org.mockito.*;

@RunWith(Parameterized.class)
public final class SmokeTest {
  private static final String FEATURE_INPUT_FOLDER = "input";
  private static final String FEATURE_OUTPUT_FOLDER = "output";
  private static final List<String> GENERATOR_NAMES = GeneratorSuite.generatorShortNames();
  private static final String RESOURCE_PREFIX = "smoke";

  private final Transpiler transpiler =
      spy(new Transpiler(OptionReader.TranspilerOptions.builder().build()));
  private final File featureDirectory;
  private final String generatorName;
  private final List<GeneratedFile> results = new LinkedList<>();

  @Before
  public void setUp() {
    Mockito.doAnswer(
            invocation -> {
              @SuppressWarnings("unchecked")
              List<GeneratedFile> generatedFiles =
                  (List<GeneratedFile>) invocation.getArguments()[0];
              results.addAll(generatedFiles);
              return true;
            })
        .when(transpiler)
        .output(any());
  }

  @Parameters(name = "{2}, {1}")
  public static Collection<Object[]> data() {
    URL smokeResourcesUrl = ClassLoader.getSystemClassLoader().getResource(RESOURCE_PREFIX);
    if (smokeResourcesUrl == null) {
      fail("Smoke test resources directory not found");
      return Collections.emptyList();
    }

    File smokeResourcesDirectory;
    try {
      smokeResourcesDirectory = new File(smokeResourcesUrl.toURI());
    } catch (URISyntaxException e) {
      fail("Unable to load smoke test resources");
      return Collections.emptyList();
    }

    File[] featureDirectoryResources = smokeResourcesDirectory.listFiles();
    if (featureDirectoryResources == null) {
      fail("No smoke test features were found");
      return Collections.emptyList();
    }
    return Arrays.stream(featureDirectoryResources)
        .filter(File::isDirectory)
        .flatMap(
            directory ->
                GENERATOR_NAMES
                    .stream()
                    .map(
                        generatorName ->
                            new Object[] {
                              directory,
                              generatorName,
                              getFeatureName(smokeResourcesDirectory, directory)
                            }))
        .collect(Collectors.toList());
  }

  public SmokeTest(
      final File featureDirectory,
      final String generatorName,
      @SuppressWarnings("unused") final String featureName) {
    this.featureDirectory = featureDirectory;
    this.generatorName = generatorName;
  }

  @Test
  public void smokeTest() {
    File inputDirectory = new File(featureDirectory, FEATURE_INPUT_FOLDER);
    File outputDirectory = new File(featureDirectory, FEATURE_OUTPUT_FOLDER);
    File outputForGeneratorDirectory = new File(outputDirectory, generatorName);

    Collection<File> referenceFiles = listFilesRecursively(outputForGeneratorDirectory);
    assumeFalse("No reference files were found", referenceFiles.isEmpty());

    assertTrue(
        transpiler.executeGenerator(
            generatorName, Collections.singletonList(inputDirectory), new HashMap<>()));

    Map<String, String> generatedContents =
        results
            .stream()
            .collect(
                Collectors.toMap(
                    generatedFile -> generatedFile.targetFile.getPath(),
                    generatedFile -> generatedFile.content));

    for (final File referenceFile : referenceFiles) {
      String relativePath = getRelativePath(outputDirectory, referenceFile);
      String generatedContent = generatedContents.get(relativePath);
      assertNotNull("File was not generated: " + relativePath, generatedContent);

      String expected = Files.readFileIntoString(referenceFile.getPath());
      assertEquals(
          "File content differs for file: " + relativePath,
          ignoreWhitespace(expected),
          ignoreWhitespace(generatedContent));
    }
  }

  private static Collection<File> listFilesRecursively(final File directory) {

    final File[] files = directory.listFiles();
    if (files == null) {
      return Collections.emptyList();
    }

    Collection<File> result = new LinkedList<>();
    for (final File file : files) {
      if (file.isDirectory()) {
        result.addAll(listFilesRecursively(file));
      } else {
        result.add(file);
      }
    }

    return result;
  }

  private static String getRelativePath(final File directory, final File file) {
    return directory.toURI().relativize(file.toURI()).getPath();
  }

  private static String ignoreWhitespace(String text) {
    return text.replaceAll("( *\n)+", "\n") // ignore repeating empty lines
        .trim(); // ignore leading and trailing whitespace
  }

  private static String getFeatureName(final File parentDirectory, final File featureDirectory) {
    return getRelativePath(parentDirectory, featureDirectory).replace("/", "");
  }
}

File: gluecodium/src/test/java/com/here/gluecodium/SmokeTest.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.gluecodium

import com.here.gluecodium.cli.OptionReader
import com.here.gluecodium.generator.common.GeneratedFile
import com.here.gluecodium.generator.common.Generator
import com.here.gluecodium.generator.common.GeneratorOptions
import com.here.gluecodium.model.lime.LimeModelLoader
import com.here.gluecodium.test.NiceErrorCollector
import io.mockk.every
import io.mockk.spyk
import junit.framework.TestCase
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Assume.assumeFalse
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import java.io.File
import java.net.URISyntaxException
import java.util.HashMap

@RunWith(Parameterized::class)
class SmokeTest(
    private val featureDirectory: File,
    private val generatorName: String,
    @Suppress("UNUSED_PARAMETER") featureName: String
) {
    @JvmField
    @Rule
    val errorCollector = NiceErrorCollector()

    private lateinit var gluecodium: Gluecodium

    private val results = mutableListOf<GeneratedFile>()

    private fun getOptions(): Pair<GluecodiumOptions, GeneratorOptions> {
        val inputDirectory = File(featureDirectory, FEATURE_INPUT_FOLDER)
        val auxDirectory = File(featureDirectory, FEATURE_AUX_FOLDER)
        val commandLineOptions = File(inputDirectory, "commandlineoptions.txt")
        if (commandLineOptions.exists()) {
            // read command line options and replace INPUT_FOLDER with absolute input path from test
            val commands = commandLineOptions.readText()
                .replace("\$INPUT_FOLDER", inputDirectory.toString())
                .replace("\$AUX_FOLDER", auxDirectory.toString())
                .split("\\s".toRegex())
            val options = OptionReader.read(commands.toTypedArray())
            TestCase.assertNotNull("Failed to read commandlineoptions.txt", options)
            return options!!
        }

        return Pair(GluecodiumOptions(), TEST_OPTIONS)
    }

    @Before
    fun setUp() {
        val options = getOptions()
        gluecodium = spyk(Gluecodium(options.first, options.second))
        every { gluecodium.output(any(), any()) }.answers {
            @Suppress("UNCHECKED_CAST")
            results.addAll(it.invocation.args[1] as List<GeneratedFile>)
            true
        }
    }

    @Test
    fun smokeTest() {
        val inputDirectory = File(featureDirectory, FEATURE_INPUT_FOLDER)
        val auxDirectory = File(featureDirectory, FEATURE_AUX_FOLDER)
        val outputDirectory = File(featureDirectory, FEATURE_OUTPUT_FOLDER)
        val validationShouldFail = File(inputDirectory, "validationfail.txt").exists()

        val generatorDirectories = listOf(generatorName) + (ADDITIONAL_GENERATOR_DIRS[generatorName] ?: emptyList())
        val referenceFiles = generatorDirectories
            .map { generatorDirectoryName -> File(outputDirectory, generatorDirectoryName) }
            .flatMap { listFilesRecursively(it) }
            .filterNot { it.name.toLowerCase().startsWith(IGNORE_PREFIX) }

        assumeFalse("No reference files were found", referenceFiles.isEmpty())

        val limeModel = LOADER.loadModel(listOf(inputDirectory.toString()), listOf(auxDirectory.toString()))
        val validationResult = gluecodium.validateModel(limeModel)
        if (validationShouldFail) {
            assertFalse(validationResult)
            return
        } else {
            assertTrue(validationResult)
        }
        assertTrue(gluecodium.executeGenerator(generatorName, limeModel, HashMap()))

        val generatedContents = results.associateBy({ it.targetFile.path }, { it.content })

        referenceFiles.forEach {
            val relativePath = getRelativePath(outputDirectory, it)
            val generatedContent = generatedContents[relativePath]
            errorCollector.checkNotNull(
                "File was not generated: $relativePath, generated files:\n${generatedContents.keys.joinToString("\n")}",
                generatedContent
            )

            if (generatedContent != null) {
                val expected = it.readText()
                errorCollector.checkEquals(
                    "File content differs for file: $relativePath",
                    ignoreUnimportantDifferences(expected),
                    ignoreUnimportantDifferences(generatedContent)
                )
            }
        }
    }

    companion object {
        private val TEST_OPTIONS = GeneratorOptions(
            cppInternalNamespace = listOf("gluecodium"),
            internalPrefix = "foobar_",
            javaNonNullAnnotation = Pair("NonNull", listOf("android", "support", "annotation")),
            javaNullableAnnotation = Pair("Nullable", listOf("android", "support", "annotation"))
        )

        private const val FEATURE_INPUT_FOLDER = "input"
        private const val FEATURE_AUX_FOLDER = "auxiliary"
        private const val FEATURE_OUTPUT_FOLDER = "output"
        private const val IGNORE_PREFIX = "ignore"
        private val ADDITIONAL_GENERATOR_DIRS = mapOf("swift" to listOf("cbridge", "cbridge_internal"))

        private val LOADER = LimeModelLoader.getLoaders().first()

        private const val RESOURCE_PREFIX = "smoke"

        @JvmStatic
        @Parameterized.Parameters(name = "{2}, {1}")
        fun data(): Collection<Array<Any>> {
            val testResourcesUrl = ClassLoader.getSystemClassLoader().getResource(RESOURCE_PREFIX)
            if (testResourcesUrl == null) {
                fail("Test resources directory not found")
                return emptyList()
            }

            val testResourcesDirectory: File
            try {
                testResourcesDirectory = File(testResourcesUrl.toURI())
            } catch (e: URISyntaxException) {
                fail("Unable to load test resources")
                return emptyList()
            }

            val featureDirectoryResources = testResourcesDirectory.listFiles()
            if (featureDirectoryResources == null) {
                fail("No test features were found")
                return emptyList()
            }
            val generatorNames = Generator.allGeneratorShortNames
            return featureDirectoryResources.filter { it.isDirectory }.sorted()
                .flatMap { directory ->
                    generatorNames.map { generatorName ->
                        arrayOf(
                            directory,
                            generatorName,
                            getFeatureName(testResourcesDirectory, directory)
                        )
                    }
                }
        }

        private fun listFilesRecursively(directory: File): Collection<File> {
            val files = directory.listFiles() ?: return emptyList()
            val result = mutableListOf<File>()
            files.forEach {
                if (it.isDirectory) {
                    result.addAll(listFilesRecursively(it))
                } else {
                    result.add(it)
                }
            }
            return result
        }

        private fun getRelativePath(directory: File, file: File) =
            directory.toPath().relativize(file.toPath()).toString()

        private fun ignoreUnimportantDifferences(text: String) =
            text
                .replace("( *[\r\n])+".toRegex(), "\n") // ignore repeating empty lines
                .trim { it <= ' ' } // ignore leading and trailing whitespace
                // treat Windows path separators as if they were Unix ones
                .split('\n').joinToString("\n") { if (it.startsWith("#include")) it.replace('/', '\\') else it }

        private fun getFeatureName(parentDirectory: File, featureDirectory: File) =
            getRelativePath(parentDirectory, featureDirectory).replace("/", "")
    }
}
